### PR TITLE
[no squash]  Check for indent spaces

### PR DESCRIPTION
**Pull request by sfence**
Continue of #14983

This PR checks the indent in C/C++ and Lua files if somewhere spaces are used.
It ignores spaces in multiline comments.

## To do

This PR is a Ready for Review

- [x] Fix found bad indents

## How to test

Run grep commands and Python script over the test file:
[test_C_file.txt](https://github.com/user-attachments/files/16647602/test_C_file.txt)

### CODE REVIEW

**Comment by cx384** [comment]
I'm not sure about disallowing less than 4 spaces after a tab, i.e. in my opinion `grep -n -P '^\s*([ ]\t|[ ]{4})'` would be strict enough. Also, I think this shouldn't get too complex, a vague check should be sufficient, otherwise you could directly use a code formatter like GNU indent or similar.
But let's wait for opinions from other code devs.

File: src/unittest/test_noise.cpp [hunk 1]
```diff
@@ -182,10 +182,10 @@ const float TestNoise::expected_2d_results[10 * 10] = {
 	15.53764, 13.83844, 12.77139, 12.73648, 13.95632, 14.97904, 15.81829,
 	18.37694, 19.73759, 13.19182, 12.71924, 11.34560, 10.78025, 11.18980,
 	12.52303, 13.45012, 14.30001, 17.43298, 19.15244, 10.93217, 10.48625,
-	 9.30923,  9.18632, 10.16251, 12.11264, 13.19697, 13.80801, 16.39567,
+	9.30923 ,  9.18632, 10.16251, 12.11264, 13.19697, 13.80801, 16.39567,
```
Comment (cx384) [inline src/unittest/test_noise.cpp:185]: e.g. this may be a controversial change

File: builtin/game/voxelarea.lua [hunk 2]
```diff
@@ -46,16 +46,16 @@ end
 function VoxelArea:index(x, y, z)
 	local MinEdge = self.MinEdge
 	local i = (z - MinEdge.z) * self.zstride +
-			  (y - MinEdge.y) * self.ystride +
-			  (x - MinEdge.x) + 1
+			(y - MinEdge.y) * self.ystride +
+			(x - MinEdge.x) + 1
```
Comment (cx384) [inline builtin/game/voxelarea.lua:50]: Same here, it is no longer aligned.

File: builtin/game/voxelarea.lua [hunk 3]
```diff
@@ -78,15 +78,15 @@ end
 function VoxelArea:contains(x, y, z)
 	local MaxEdge, MinEdge = self.MaxEdge, self.MinEdge
 	return (x >= MinEdge.x) and (x <= MaxEdge.x) and
-		   (y >= MinEdge.y) and (y <= MaxEdge.y) and
-		   (z >= MinEdge.z) and (z <= MaxEdge.z)
+			(y >= MinEdge.y) and (y <= MaxEdge.y) and
```
Comment (cx384) [inline builtin/game/voxelarea.lua:81]: Same here, it is no longer aligned.

File: builtin/game/voxelarea.lua [hunk 4]
```diff
@@ -46,16 +46,16 @@ end
...
+			(x - MinEdge.x) + 1
 	return math_floor(i)
 end
 
 function VoxelArea:indexp(p)
 	local MinEdge = self.MinEdge
 	local i = (p.z - MinEdge.z) * self.zstride +
-			  (p.y - MinEdge.y) * self.ystride +
-			  (p.x - MinEdge.x) + 1
+			(p.y - MinEdge.y) * self.ystride +
```
Comment (cx384) [inline builtin/game/voxelarea.lua:57]: Same here, it is no longer aligned.

File: builtin/game/voxelarea.lua [hunk 5]
```diff
@@ -46,16 +46,16 @@ end
 function VoxelArea:index(x, y, z)
 	local MinEdge = self.MinEdge
 	local i = (z - MinEdge.z) * self.zstride +
-			  (y - MinEdge.y) * self.ystride +
-			  (x - MinEdge.x) + 1
```
Comment (cx384) [inline builtin/game/voxelarea.lua:50]: Same here, it is no longer aligned.

File: builtin/game/voxelarea.lua [hunk 6]
```diff
@@ -46,16 +46,16 @@ end
...
+			(y - MinEdge.y) * self.ystride +
```
Comment (cx384) [inline builtin/game/voxelarea.lua:49]: Same here, it is no longer aligned.

Comment (cx384) [inline builtin/game/voxelarea.lua:49]: Same here, it's no longer aligned.

**Comment by Desour** [comment]
There's a surprisingly high amount of space indentation.

But also half of the changes are getting rid of valid alignment. Alignment is too useful for readability to forbid imo. Which is why I'm against this PR in its current state. :-1: 

Something more conservative could maybe work, i.e.:
* No tabs after any non-tab character.
* If a line starts with k tabs, and the next line has k' < k tabs, there must be no space character before any non-whitespace character.
  (Rationale: Alignment never takes away tabs. Hence this indicates space being used for indentation.)

**Comment by sfence** [comment]
@cx384 I tried to catch only more than 3 spaces after the tab, but it still does not work well.

So I used the less restrictive solution mentioned by @Desour.

Some badly used spaces are not recognized now.

Do you think this solution is better?

File: src/client/content_cso.cpp [hunk 7]
```diff
@@ -55,7 +55,7 @@ class SmokePuffCSO: public ClientSimpleObject
 		MapNode n = env->getMap().getNode(floatToInt(pos, BS), &pos_ok);
 		light = pos_ok ? decode_light(n.getLightBlend(env->getDayNightRatio(),
 							env->getGameDef()->ndef()->getLightingFlags(n)))
-		               : 64;
+							: 64;
```
Comment (cx384) [inline src/client/content_cso.cpp:58]:
```suggestion
						: 64;
```
IMO, the previous alignment was better, since the line above is part of the function call two lines above.
Alternately, for compactness, `: 64;` could be put directly at the end of the line above, or the ternary operator could be split properly, but then it would get too big.

(Also, according to our code style guildlines we only add two tabs for those types of line breaks.)

File: src/client/game.cpp [hunk 8]
```diff
@@ -4089,8 +4089,8 @@ void Game::updateFrame(ProfilerGraph *graph, RunStats *stats, f32 dtime,
 		float old_brightness = sky->getBrightness();
 		direct_brightness = client->getEnv().getClientMap()
 				.getBackgroundBrightness(MYMIN(runData.fog_range * 1.2, 60 * BS),
-					daynight_ratio, (int)(old_brightness * 255.5), &sunlight_seen)
-				    / 255.0;
+						daynight_ratio, (int)(old_brightness * 255.5), &sunlight_seen)
+				/ 255.0;
```
Comment (cx384) [inline src/client/game.cpp:4093]:
```suggestion
					daynight_ratio, (int)(old_brightness * 255.5), &sunlight_seen)
					/ 255.0;
```
You removed 4 spaces, but placed the tab at the wrong line

File: src/script/common/c_content.h [hunk 9]
```diff
@@ -77,130 +77,130 @@ extern struct EnumString es_TileAnimationType[];
 
 extern const std::array<const char *, 33> object_property_keys;
 
-void               read_content_features     (lua_State *L, ContentFeatures &f,
-                                              int index);
-void               push_content_features     (lua_State *L,
-                                              const ContentFeatures &c);
+void read_content_features (lua_State *L, ContentFeatures &f,
+		int index);
```
Comment (cx384) [inline src/script/common/c_content.h:81]: Just turn those into one line if it gets this small, everywhere in this file.

However, it would be good to first get opinions from other core devs before removing those alignments.
I personally don't really care, but I guess it makes sense for consistency.

File: src/script/common/c_converter.h
Comment (cx384) [inline src/script/common/c_converter.h:1]: Same as in src/script/common/c_content.h

**Review by cx384 — COMMENTED** [review]
Otherwise, I'm fine with the indentions, but I haven't looked at the script yet.

File: src/mapgen/treegen.cpp [hunk 10]
```diff
@@ -292,29 +292,29 @@ treegen::error make_ltree(MMVManip &vmanip, v3s16 p0,
...
+	  +  - yaw the turtle right by angle degrees
+	  -  - yaw the turtle left by angle degrees
+	  &  - pitch the turtle down by angle degrees
+	  ^  - pitch the turtle up by angle degrees
+	  /  - roll the turtle to the right by angle degrees
+	  *  - roll the turtle to the left by angle degrees
+	  [  - save in stack current state info
+	  ]  - recover from stack state info
+
+	 */
```
Comment (cx384) [inline src/mapgen/treegen.cpp:317]: I don't see any benefits of this change.

Comment (sfence) [reply src/mapgen/treegen.cpp:317]: This was a situation when the beginning of a multiline comment was indented by one tabulator, and the rest of it was indented by spaces.
The check process is sensitive to this now. So it was reported as an error.

If it is a problem, I can ignore multiple comments and ignore their indents.

File: src/client/content_cso.cpp (hunk 7 above)
Comment (sfence) [reply src/client/content_cso.cpp:58]: I will have to add some style of exceptions if we want indents to be checked and we want to allow exceptions at the same time.

But in that case, the list of exceptions has to be kept up to date.

File: src/client/game.cpp (hunk 8 above)
Comment (sfence) [reply src/client/game.cpp:4093]: I have indented 4092 by two tabs because it is part of the method getBackgroundBrightness calling from the previous line.
and line 4093 is not part of it. So I continue in the same indent as 4091 did.

If I understand well from our code style guide, if a command is not finished on one line, the next line should be two tabs indented from the line where the command begins.

File: src/client/content_cso.cpp (hunk 7 above)
Comment (cx384) [reply src/client/content_cso.cpp:58]: > IMO, the previous alignment was better ...

I meant the alignment not the indents.
You replaced 15 spaces with 5 tabs.
I want it to be 4 tabs instead.

I don't get why this should be an exception.

File: src/client/game.cpp (hunk 8 above)
Comment (cx384) [reply src/client/game.cpp:4093]: Makes sense, I thought, you only replaced the spaces with a tab.

File: src/mapgen/treegen.cpp (hunk 10 above)
Comment (sfence) [reply src/mapgen/treegen.cpp:317]: But disabling it means that indents like here [`src/util/string.h`](https://github.com/minetest/minetest/pull/15002/files/aa19202aad794b9c6ee9520d7cb75c504eb18da5#diff-71badc1cc71ba46244f7841a088251bb294265f4fe9e662c0ad6b15be540eee4) and  here [`src/unittest/test_map_settings_manager.cpp`](https://github.com/minetest/minetest/pull/15002/files/aa19202aad794b9c6ee9520d7cb75c504eb18da5#diff-3fd3459a4043212e05ed20dc3563b4f114f4e48b9d7b58dcbc84b0bf36efc83a) will not be recognized as errors.

File: src/client/content_cso.cpp (hunk 7 above)
Comment (sfence) [reply src/client/content_cso.cpp:58]: Yes, I overindent it. 

My comment was related to "IMO, the previous alignment was better, since the line ..."

File: src/mapgen/treegen.cpp (hunk 10 above)
Comment (cx384) [reply src/mapgen/treegen.cpp:317]: Hm, I'm fine with both, but I guess it would make sense to allow spaces here, like for the big comment in src/server/clientiface.h

Comment (sfence) [reply src/mapgen/treegen.cpp:317]: In `clientiface.h` it is not a problem because the first `/*` is not indented.

So, I will keep it for now, if somebody else disagrees with this solution, I change it here and in the script.

**Comment by Desour** [comment]
Could you please add some comments, so that for example one doesn't have to read through the whole preprocess script to find out what it does?

So, what this disallows now is spaces after tabs and multiline comments that start in the same line that another multiline comment ends?
Why do we need latter? And wouldn't former still break alignment after indentation?

**Comment by sfence** [comment]
> Could you please add some comments, so that for example one doesn't have to read through the whole preprocess script to find out what it does?

I added some comments, I hope it helps.

> So, what this disallows now is spaces after tabs and multiline comments that start in the same line that another multiline comment ends? Why do we need latter? And wouldn't former still break alignment after indentation?

Now spaces at the beginning of the line and spaces after tabs only if the indent is smaller than on the previous line.
In other cases, spaces after the indent is ignored.

The multiline comments part is questionable.
The idea is to detect something like this:
```
	{
		soem code
		/*
 wtf
  wtf
		*/
	}
```

File: .github/workflows/whitespace_checks.yml [hunk 11]
```diff
@@ -34,6 +36,32 @@ jobs:
...
+          if git ls-files | grep -E '\.lua$' | xargs grep -n '\]\]--.*--\[\['; then echo -e "\033[0;31mUnsupported combination of multiline comments. New multiline comment should begin on new line."; (exit 1); else (exit 0); fi
+      # This prepare files for final check 
+      # See python script ./util/ci/indent_tab_preprocess.py for details.
+      - name: Preprocess files
+        run: |
+          git ls-files | grep -E '^src/.*\.cpp$|^src/.*\.[ch]$' | xargs -L 1 -P $(($(nproc) + 1)) python3 ./util/ci/indent_tab_preprocess.py "/\\*" "\\*/" 2
+          git ls-files | grep -E '\.lua$' | xargs -L 1 -P $(($(nproc) + 1))  python3 ./util/ci/indent_tab_preprocess.py "--\\[\\[*" "--\\]\\]" 4
+      # Check for bad indent.
+      # This runs over preprocessed files.
+      # If there is any space on line begining or after tab,
```
Comment (Desour) [inline .github/workflows/whitespace_checks.yml:59]:
```suggestion
      # If there is any remaining space on line beginning or after tab,
```

File: util/ci/indent_tab_preprocess.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+def main():
+	if len(sys.argv) != 5:
+		#print("Bad arguments")
+		#print(sys.argv)
+		exit()
+	# multiline comment begining and end from command line
+	mlc_begin = sys.argv[1]
+	mlc_end = sys.argv[2]
+	mlc_end_len = int(sys.argv[3])
```
Comment (Desour) [inline util/ci/indent_tab_preprocess.py:69]:
```suggestion
	mlc_end_len = len(mlc_end)
```

File: util/ci/indent_tab_preprocess.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+	for line in file:
+		if limit == -1:
+			find = re.search(re_mlc_begin, line)
+			if find:
+				if not re.search(re_mlc_begin_end, line):
+					# enter in block comment mode
+					find = re.search("^\\s*", line)
+					limit = find.span()[1]
+			find = re.search("^\\t*", line)
+			end = find.span()[1]
```
Comment (Desour) [inline util/ci/indent_tab_preprocess.py:94]:
```suggestion
			end = find.end()
```

File: util/ci/indent_tab_preprocess.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+	#print(re_mlc_end)
+	#print(file)
+
+	limit = -1
+	prev = 80
+	for line in file:
+		if limit == -1:
+			find = re.search(re_mlc_begin, line)
+			if find:
+				if not re.search(re_mlc_begin_end, line):
```
Comment (Desour) [inline util/ci/indent_tab_preprocess.py:89]: I don't think you can find a begin-end pair if you didn't find a begin.

File: util/ci/indent_tab_preprocess.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+		exit()
+	# multiline comment begining and end from command line
+	mlc_begin = sys.argv[1]
+	mlc_end = sys.argv[2]
+	mlc_end_len = int(sys.argv[3])
+	file = open(sys.argv[4], "r")
+
+	re_mlc_begin = "^.*{}".format(mlc_begin)
+	re_mlc_begin_end = "^.*{}.*{}".format(mlc_begin, mlc_end)
+	re_mlc_end = "^.*{}".format(mlc_end)
```
Comment (Desour) [inline util/ci/indent_tab_preprocess.py:74]: Wouldn't simple non-pattern string search be easier?

File: util/ci/indent_tab_preprocess.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+	lines = []
+
+	#print(re_mlc_begin)
+	#print(re_mlc_begin_end)
+	#print(re_mlc_end)
+	#print(file)
+
+	limit = -1
+	prev = 80
```
Comment (Desour) [inline util/ci/indent_tab_preprocess.py:84]: What are these? Please comment.

File: src/mapgen/treegen.cpp (hunk 10 above)
Comment (Desour) [reply src/mapgen/treegen.cpp:317]: Imo new version makes sense.

File: src/script/common/c_content.h [hunk 17]
```diff
@@ -77,130 +77,99 @@ extern struct EnumString es_TileAnimationType[];
 
 extern const std::array<const char *, 33> object_property_keys;
 
-void               read_content_features     (lua_State *L, ContentFeatures &f,
-                                              int index);
-void               push_content_features     (lua_State *L,
-                                              const ContentFeatures &c);
+void read_content_features (lua_State *L, ContentFeatures &f, int index);
```
Comment (Desour) [inline src/script/common/c_content.h:80]:
```suggestion
void read_content_features(lua_State *L, ContentFeatures &f, int index);
```
The space was just there for the alignment. If we remove the alignment (which is fine imo), don't keep the one space.
(Also everywhere below.)

File: .github/workflows/whitespace_checks.yml [hunk 18]
```diff
@@ -34,6 +36,32 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      # Line endings are already ensured by .gitattributes
+      # Multiple multline comments is not supported by this cehck 
+      # and there is no reason to use them
+      # So lines like: "/* */ /*" or "*/ a = 5; /*" will result in error
+      - name: Check for unsupported multiline comments
+        run: |
+          if git ls-files | grep -E '^src/.*\.cpp$|^src/.*\.[ch]$' | xargs grep -n '\*/.*/\*'; then echo -e "\033[0;31mUnsupported combination of multiline comments. New multiline comment should begin on new line."; (exit 1); else (exit 0); fi
```
Comment (Desour) [inline .github/workflows/whitespace_checks.yml:49]: Are line breaks possible btw.? These lines are so long.

**Review by Desour — COMMENTED** [review]
Python script is a bit hard to follow.

Whitespace changes look mostly good.

File: util/ci/indent_tab_preprocess.py (hunk 12 above)
Comment (sfence) [reply util/ci/indent_tab_preprocess.py:69]: It can't be done. The regex comment line string does not have the same length as the non-regex representation.

File: util/ci/indent_tab_preprocess.py (hunk 14 above)
Comment (sfence) [reply util/ci/indent_tab_preprocess.py:89]: `if find:` condition should be true only if the beginning is found.

File: util/ci/indent_tab_preprocess.py (hunk 15 above)
Comment (sfence) [reply util/ci/indent_tab_preprocess.py:74]: For begin and end, probably yes. For the `begin_end `combination, I do not thing so.

File: .github/workflows/whitespace_checks.yml [hunk 19]
```diff
@@ -32,14 +34,85 @@ jobs:
...
+            xargs grep -n '\*/.*/\*';\
+          then
+            echo -e "\033[0;31mUnsupported combination of multiline comments. New multiline comment should begin on new line.";\
+            (exit 1);\
+          else\
+            (exit 0);\
+          fi
+          if git ls-files |\
+            grep -E '\.lua$' |\
+            xargs grep -n -e '--\]\].*--\[\[';\
```
Comment (Desour) [inline .github/workflows/whitespace_checks.yml:69]: Also not `--]]`. It's just `]]`.

```suggestion
            xargs grep -n -e '\]\].*--\[\[';\
```

(If you see `--]]` being used somewhere, that's because someone wants the end to become a single line comment when they remove the starter. FYI.)

File: util/ci/indent_tab_preprocess.py (hunk 12 above)
Comment (Desour) [reply util/ci/indent_tab_preprocess.py:69]: Can you use `re.escape()`?

File: util/ci/indent_tab_preprocess.py (hunk 15 above)
Comment (Desour) [reply util/ci/indent_tab_preprocess.py:74]: You'd first search for the begin, then for the end, with the offset. But whatever.

File: util/ci/indent_tab_preprocess.py (hunk 14 above)
Comment (Desour) [reply util/ci/indent_tab_preprocess.py:92]: Sorry, got an off-by-1 error for bools in my brain.

File: util/ci/indent_tab_preprocess.py [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+	#print(re_mlc_begin)
+	#print(re_mlc_begin_end)
+	#print(re_mlc_end)
+	#print(file)
+
+	# -1 means not in the multiline comment
+	# other values mean multiline comment indent in the number of tabs
+	limit = -1
+	# this value stores thr previous line indent in code mode (not in multiline comment)
```
Comment (Desour) [inline util/ci/indent_tab_preprocess.py:86]:
```suggestion
	# this value stores the previous line indent in code mode (not in multiline comment)
```

File: util/ci/indent_tab_preprocess.py (hunk 13 above)
Comment (Desour) [reply util/ci/indent_tab_preprocess.py]: Was not the only place.

**Comment by Desour** [comment]
(See also unresolved older comments.)

It's just a CI script, so it doesn't need to be perfect, right? I don't want to be overly pedantic.

File: util/ci/indent_tab_preprocess.py [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+	prev = 80
```
[the remark this answers is not in the document]
Comment (sfence) [reply util/ci/indent_tab_preprocess.py:87]: It will cause the spaces after the tabs on the first line to be ignored.
So errors like in [src/unittest/test_random.cpp](https://github.com/minetest/minetest/pull/15002/files#diff-79e35cbfe7408d8e21e7f9bc27ef68e49b940299dd80c619e6a37619fcfc5dc3) will not be reported;

First line:
```
/* block comment on file beginning without space indent
 /* block comment on file beginning with space indent
```

File: .github/workflows/whitespace_checks.yml (hunk 19 above)
Comment (sfence) [reply .github/workflows/whitespace_checks.yml]: I forgot about this totally. :)

**Comment by sfence** [comment]
@Desour Should be fixed now.

**Review by Desour — APPROVED** [review]
Looks fine from what I can tell.

One trivial thing I noticed: The python script is not marked executable. The other files in util/ci are.

Btw. add [no squash] to your PR title if you want no squashing.

**Comment by Zughy** [comment]
@sfence rebase needed

**Comment by Desour** [comment]
(My approval holds.)
